csharp/ql/src/Security Features/CWE-022/TaintedPath.ql

```diff
@@ -14,7 +14,9 @@
  */
 import csharp
 import semmle.code.csharp.security.dataflow.TaintedPath::TaintedPath
+import semmle.code.csharp.dataflow.DataFlow::DataFlow::PathGraph
 
 from TaintTrackingConfiguration c, Source source, Sink sink
 where c.hasFlow(source, sink)
-select sink, "$@ flows to here and is used in a path.", source, "User-provided value"
+select sink, source.getPathNode(c), sink.getPathNode(c),
+  "$@ flows to here and is used in a path.", source, "User-provided value"
```

Comment:
I think I would prefer
```
from TaintTrackingConfiguration c, DataFlow::PathNode source, DataFlow::PathNode sink
where c.hasFlowPath(source, sink)
select sink, source, sink, "$@ flows to here and is used in a path.", source, "User-provided value"
```

Reply:
Agree.

Reply:
Or actually I think this is slightly better:
```
from TaintTrackingConfiguration c, DataFlow::PathNode source, DataFlow::PathNode sink
where c.hasFlowPath(source, sink)
select sink.getNode(), source, sink, "$@ flows to here and is used in a path.", source.getNode(), "User-provided value"
```

Reply:
`toString()` is different when you have field flow, as the `PathNode` will include the accesspath.

Reply:
Ah, makes sense.

Reply:
Actually, if you look at the test output, `.getNode()` doesn't seem to make a difference, so I would be inclined to leave it as it is.

Reply:
That is only because C# doesn't yet have field flow.

Reply:
Ok. But we don't often use the raw `toString()` text in messages anyway, and in the result object, we don't render the `toString` either. So I don't think we need to universally change things to `.getNode()` yet.

Reply:
Another problem with using `PathNode`s is that there can in general be multiple path nodes that map to the same dataflow node (for example, nodes with different call contexts). While we want to report all different paths, we only want one query result for each source/sink data flow node.

Reply:
It seems like there's a choice between
```
from Configuration c, DataFlow::Node source, DataFlow::Node sink
where c.hasFlow(source, sink)
select sink, source.getPathNode(c), sink.getPathNode(c),
  "$@ could flow here.", source, "User input"
```
vs
```
from Configuration c, DataFlow::PathNode source, DataFlow::PathNode sink
where c.hasFlowPath(source, sink)
select sink.getNode(), source, sink,
  "$@ could flow here.", source.getNode(), "User input"
```
The second is actually longer. Either way, this mapping to and fro seems somewhat unsatisfactory. I have a slight preference for the first version actually, because

- The query is actually dealing with dataflow nodes. Paths are an implementation detail.
- You often use a specific type of source and sink, not `DataFlow::Node`. Often the source and sink have additional predicates relating to information about the source and the sink. `PathNodes` don't capture any of this.
- You could forget to call `getNode()`.

Anyway, this sounds like something to be discussed offline in the CPH office.

Reply:
I think those are some convincing arguments, @calum!

Reply:
You're missing an important argument for using `PathNode`.  There are potentially many `PathNode`s corresponding to a single `Node` (e.g. if there are multiple configurations in scope with overlapping sources or sinks, which can easily happen), and in that case using `getPathNode` to get back to the `PathNode`s becomes semantically shaky, as you'll risk multiplying your results.

Reply:
Yes but `getPathNode(c)` ensures that only nodes of configuration `c` are used in the path. Therefore I think the risk of multiplication -- which only occurs in the `edges` relation -- is the same in either case.

Reply:
Still, you may get a pair of path nodes for which there isn't actually flow, as guaranteed by `c.hasFlowPath()`. So, unless the engine filters out results in the `select` predicate, for which the `edges` predicate does not produce a path from the `select` predicate's second to third column, I think we need to use the other approach.